src/shogun/kernel/MaternKernel.cpp

```diff
@@ -0,0 +1,197 @@
+		        ? std::numeric_limits<float32_t>::epsilon()
+		        : dist;
+		float64_t ratio = std::sqrt(2 * m_nu) * adjusted_dist / m_width;
+		// bessel function of the second kind
+#ifdef __clang__
+		error(
+		    "Shogun compiled on clang does not support Matern kernels with "
+		    "width other than 0.5, 1.5 and 2.5. The current width is {}",
+		    m_width);
+#else
+		float64_t bessel = std::cyl_bessel_k(m_nu, ratio);
+#endif
+		result = (std::pow(2, 1 - m_nu) / std::tgamma(m_nu)) *
+		         std::pow(ratio, m_nu) * bessel;
+	}
+
+	return result;
+}
+
+SGMatrix<float64_t> MaternKernel::get_parameter_gradient(
+    Parameters::const_reference param, index_t index)
+{
+	require(lhs, "Left hand side features must be set!");
+	require(rhs, "Right hand side features must be set!");
+
+	if (param.first == "width")
+	{
+		SGMatrix<float64_t> derivative = SGMatrix<float64_t>(num_lhs, num_rhs);
+		std::function<float64_t(float64_t)> gradient_func;
+
+		// the gradients Matern wrt m_width were computed with WolframAlpha
+		if (Math::fequals(m_nu, 0.5, std::numeric_limits<float64_t>::epsilon()))
+		{
+			const float64_t m_width_squared = std::pow(m_width, 2);
+			gradient_func = [&](const float64_t& dist) {
+				return (dist * std::exp(-dist / m_width)) / m_width_squared;
+			};
+		}
+		else if (Math::fequals(
+		             m_nu, 1.5, std::numeric_limits<float64_t>::epsilon()))
+		{
+			const float64_t m_width_cubed = std::pow(m_width, 3);
+			gradient_func = [&](const float64_t& dist) {
+				return (3 * std::pow(dist, 2) *
+				        std::exp(-(std::sqrt(3) * dist) / m_width)) /
+				       m_width_cubed;
+			};
+		}
+		else if (Math::fequals(
+		             m_nu, 2.5, std::numeric_limits<float64_t>::epsilon()))
+		{
+			const float64_t m_width_power_4 = std::pow(m_width, 4);
+			gradient_func = [&](const float64_t& dist) {
+				return (5 * std::pow(dist, 2) *
+				        std::exp(-(std::sqrt(5) * dist) / m_width) *
+				        (m_width + std::sqrt(5) * dist)) /
+				       (3 * m_width_power_4);
+			};
+		}
+		else
+		{
+#ifdef __clang__
+			error(
+			    "Shogun compiled on clang does not support Matern kernels with "
+			    "width other than 0.5, 1.5 and 2.5. The current width is {}",
+			    m_width);
+#endif
+			gradient_func = [&](const float64_t& dist) {
+				constexpr float64_t epsilon = 1E-6;
+				auto k = [&](const float64_t& width) {
+					const float64_t adjusted_dist =
+					    Math::fequals(
+					        dist, 0.0,
+					        std::numeric_limits<float64_t>::epsilon())
+					        ? std::numeric_limits<float32_t>::epsilon()
+					        : dist;
+					float64_t ratio =
+					    std::sqrt(2 * m_nu) * adjusted_dist / width;
+					float64_t bessel = std::cyl_bessel_k(m_nu, ratio);
+					return (std::pow(2, 1 - m_nu) / std::tgamma(m_nu)) *
+					       std::pow(ratio, m_nu) * bessel;
+				};
+				return (k(m_width + epsilon) - k(m_width)) / epsilon;
```

Comment:
@karlnapf of the matern kernel is a monster (I also think it is quite unstable outside the widths that are usually used, as I was getting very different result between WolframAlpha gradient and the numerical gradient. So I just put the numerical gradient (like sklearn does)

Reply:
for reference https://www.wolframalpha.com/input/?i=derivative+(2^(1+-+nu)+%2F+Gamma(nu))+*+(sqrt(2+*+nu)+*+d+%2F+w)+^+nu+*+BesselK(nu%2C+sqrt(2+*+nu)+*+d+%2F+w)+wrt+w

Reply:
ok!

Reply:
maybe add some comments in the code so that this reasoning is transparent in case someone touches it later

Reply:
sure! I think the main point is that It is more expensive to compute the gradient analytically than numerically. As for the value of the delta in the gradient calculation, it was what gave me the correct values for the width values that we compute analytically (i.e. I computed the 1/2 kernel gradient numerically and this delta gave a good result). I think it might be a bit high though?